### DISCUSSION TITLE
IP_RECVERR and Nwrite issue

### DISCUSSION THREAD

**Discussion by robertscottbeattie** [esnet/iperf#1804] [Q&A]
Hi, I am not an expert on socket programming, but I am trying to get the network error messages from the stream socket by setting:

setsockopt(s, SOL_IP, IP_RECVERR, &n, sizeof(n));

Which seems to work OK. However occasionally iperf stops working because the write call in Nwrite returns -1. When I look at the errno I see "No route to host" on the machine I am working on.

Normally this would just cause a NET_HARDERROR to be generated and iperf to stop working. If I ignore the error and just try the write again it seems to work OK.

Does anyone have any thoughts / knowledge in this area and can explain whats going on?

**Comment by davidBar-On** [comment]
Hi, I am also not a socket expert, but I believe that it would help if you will give more details about you implementation.  I.e., are you using the iperf3 application or the iperf3 library?  In either case, can you explain (and show the relevant code) how are you reading the errors in parallel to the iperf3 code? etc.

**Comment by robertscottbeattie** [comment]
I am using the iperf3 application. I added the following line to  iperf_common_sockopts:

   /* */
    if (getsockdomain(s) == AF_INET) {
        if (setsockopt(s, SOL_IP, IP_RECVERR, &n, sizeof(n)) < 0)
            printf("setsockopt IP_RECVERR. errno=%s\n", strerror(errno));
    } else if (getsockdomain(s) == AF_INET6) {
        if (setsockopt(s, SOL_IPV6, IPV6_RECVERR, &n, sizeof(n)) < 0)
            printf("setsockopt IPV6_RECVERR. errno=%s\n", strerror(errno));
    }

I modified Nwrite to this:

    while (nleft > 0) {
        r = write(fd, buf, nleft);
        if (r < 0) {
            printf("Nwrite : errno = %s\n", strerror(errno));
            continue;
        }

        nleft -= r;
        buf += r;
    }

As sometimes the return value r = -1. The error reported is "No route to host". This happens relatively infrequently and it depends on the machine. The second attempt to write works OK. I was just curious about why the write behaviour would change. If there was something I was failing to understand.

**Comment by davidBar-On** [comment]
Can you describe the architecture of your network between the client and the server?  It may be that the problem is in one of the components in the way (firewall, switch, router, etc.).  Also, is this a TCP or UDP test?  Can you show the client and server iperf3 commands (no need to disclose IP addresses, domain names, etc.).
